test/test_eager_transforms.py

```diff
@@ -1636,15 +1636,41 @@ def f(x):
         def g(x):
             return ()
 
-        def h(x):
-            return x, [x]
-
-        with self.assertRaisesRegex(RuntimeError, "a Tensor or Tensors"):
+        with self.assertRaisesRegex(RuntimeError, "Expected non-empty output"):
             jvp(f, (x,), (t,))
-        with self.assertRaisesRegex(RuntimeError, "non-empty"):
+
+        with self.assertRaisesRegex(RuntimeError, "Expected non-empty output"):
             jvp(g, (x,), (t,))
-        with self.assertRaisesRegex(RuntimeError, "a Tensor or Tensors"):
-            jvp(h, (x,), (t,))
+
+        # Check list output
+        out = jvp(lambda x: [x, x.sum()], (x,), (t,))
+        for i in range(2):
+            assert isinstance(out[i], list) and len(out[i]) == 2
+
+        # Check dict output
+        out = jvp(lambda x: {"x": x, "xsum": x.sum()}, (x,), (t,))
+        for i in range(2):
+            assert isinstance(out[i], dict) and len(out[i]) == 2 and "xsum" in out[i]
+
+        def composite_output(x):
+            out = x.sum()
+            return [
+                (out, {"a": x, "out": [x, out]}),
+                int(out),
+                out > 0,
+            ]
+
+        out = jvp(composite_output, (x,), (t,))
+        for i in range(2):
+            assert isinstance(out[i], list)
+            assert isinstance(out[i][0], tuple) and \
+                isinstance(out[i][0][1], dict)
+
+        # Checks primals and tangents for int(out)
+        assert isinstance(out[0][1], int) and isinstance(out[1][1], torch.Tensor)
+        # Checks primals and tangents for out > 0
+        assert isinstance(out[0][2], torch.Tensor) and out[0][2].dtype == torch.bool \
+            and isinstance(out[1][2], torch.Tensor)
```

Comment:
I think we should restrict jvp to only allow returns that are Tensors

Reply:
I was based on your comment from the issue: https://github.com/pytorch/functorch/issues/268#issuecomment-1009354325 and replicating JAX behaviour if possible. 
JAX does not allow to return any type of output neither, e.g. if returning a float it will raise an error, but user can return int and bool. I was trying to replicate this behaviour...

Reply:
Hmm, I wonder if we can find examples of how this is used

Reply:
Let me know if you end up finding examples where people do return int/bool from a function and then use it with the jvp transform.

functorch doesn't have parity with JAX on how scalars are handled. If the user wants to differentiate a scalar, for the other transforms (e.g. grad), we require them to use a scalar Tensor.

It should be safer for now to be consistent with our other transforms and require that the function return only Tensors -- it's easy to relax this constraint if users come to us asking to relax it. What are your thoughts?

Reply:
I'm OK with returning only tensors requirement. If we would like to expand that we may have to properly design and test it (and not in this PR).

As for int/bool support, maybe this jax PR is relevant: https://github.com/google/jax/pull/4039